File: backend/i18n.py
"""Python-side translation lookup for the few strings the backend itself
needs to display (the native tray icon menu, which has no JS runtime).
Reads the same locale JSON files the frontend uses, so there's still only
one set of translation strings in the whole app.
"""
import json

from . import paths

_cache = {}


def translate(lang, key):
    lang = (lang or "EN").lower()
    if lang not in _cache:
        try:
            with open(paths.resource_path("frontend", "locales", f"{lang}.json"), "r", encoding="utf-8") as f:
                _cache[lang] = json.load(f)
        except Exception:
            _cache[lang] = {}
    return _cache[lang].get(key, key)

File: dotnet/src/FlightOpsHub.Core/I18n.cs
using System.Text.Json.Nodes;

namespace FlightOpsHub.Core;

/// <summary>
/// Port of backend/i18n.py - translation lookup for the few strings the
/// backend itself needs (the native tray icon menu, which has no JS
/// runtime). Reads the same locale JSON files the frontend uses.
/// </summary>
public class I18n
{
    private readonly string _localesDir;
    private readonly Dictionary<string, JsonObject> _cache = new();

    public I18n(string localesDir)
    {
        _localesDir = localesDir;
    }

    public string Translate(string? lang, string key)
    {
        lang = (string.IsNullOrEmpty(lang) ? "EN" : lang).ToLowerInvariant();
        if (!_cache.TryGetValue(lang, out var strings))
        {
            try
            {
                var text = File.ReadAllText(Path.Combine(_localesDir, $"{lang}.json"));
                strings = JsonNode.Parse(text) as JsonObject ?? new JsonObject();
            }
            catch
            {
                strings = new JsonObject();
            }
            _cache[lang] = strings;
        }
        return strings[key] is JsonValue value && value.TryGetValue<string>(out var result) ? result : key;
    }
}
